### PR TITLE
Issue 547: Allow load extra lifecycle components when running a bookie server.

### PR DESCRIPTION
**Pull request by sijie**
Descriptions of the changes in this PR:

- make a common `ServerLifecycleComponet` for the components loaded by a bookie server
- add a flag to allow start extra components.
- add test cases

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/bookkeeper-precommit-pullrequest-docker/94/

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/server/component/package-info.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * The lifecycle components that run on a bookie server.
+ */
+package org.apache.bookkeeper.server.component;
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/server/component/package-info.java]: no new line ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/util/ReflectionUtils.java [hunk 2]
```diff
@@ -36,6 +35,30 @@
...
+     * @param xface class interface
+     * @return the class object associated with the class or interface with the given string name.
+     */
+    public static <U> Class<? extends U> forName(String className,
+                                                 Class<U> xface) {
+
+        // Construct the class
+        Class<?> theCls;
+        try {
+            theCls = Class.forName(className);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/util/ReflectionUtils.java]: I think is does not work very well, it is better to use Thread.currentThread().getContextClassloader()
I know that for bookie started from the command line this is not an issue, but if you embed the bookie in another process the class loader will be important.
I see that anyway the same approach is used in another method in this class.
So I am fine with this change, we can improve it later.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/server/component/package-info.java (hunk 1 above)
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/server/component/package-info.java]: these packages are under validation of current checkstyle rules. I don't think there is such one rule.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/util/ReflectionUtils.java (hunk 2 above)
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/util/ReflectionUtils.java]: I don't think we have ever used `Thread.currentThread().getContextClassLoader()`. I think that is typically useful in a multi-classloader environment. I am not sure the bookkeeper usage will ever get into this situation.

**Pull request closed without merging**